### ISSUE TITLE
OLD dataclasses and namedtuples concept exercise

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2361]
This issue describes how to implement the `dataclasses` and `namedtuples` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/creation/use of `dataclasses` and `namedtuples` in Python. 

## Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes` in Python
- understand/create/use `dataclasses` and the `@dataclass` decorator
- learn some additional methods from the `dataclass` module
  - `dataclasses.fields`
  - `dataclasses.asdict` & `dataclasses.astuple`
  - `dataclasses.replace`
  - default factory functions
- learn an immutable alternative to `dataclasses` -- `collections.namedtuple`
  - pros and cons of each method for creating data-focused `classes`

  

## Out of scope

- `class-inheritance`, `multiple-inheritance`, `__super()__`, class `mixins`
- `class-composition`  -- (_beyond the composition needed for the decorators in this exercise_)
- `comprehensions`
- `decorators` outside of `@dataclass`
- `generators`
- `coroutines`
- `descriptors` (_these will get their own exercise_)
- using a `class` as a decorator, beyond the methods made available for `dataclasses`
- `type annotaions`, beyond what is used in `dataclasses`
- `collections` module, outside of `collections.namedtuple` (_the rest will get their own exercise_)

 

## Concepts

- `classes`
- `class attributes`
- `class members`
- `dataclasses`, `@dataclass`
- `decorators`
-  `collections.namedtuples()`

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `booleans`
- `classes`
- `class-customization`
- `comparisons`
- `decorators`
- `dicts`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

  

## Resources to refer to

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Python Docs: dataclasses](https://docs.python.org/3/library/dataclasses.html#module-dataclasses)
- [Python Docs: collections.namedtuple()](https://docs.python.org/3/library/collections.html#collections.namedtuple)
- [PEP 0557](https://www.python.org/dev/peps/pep-0557/)
- [PEP 0526](https://www.python.org/dev/peps/pep-0526/)
- [Real Python: Data Classes in Python 3.7+ (Guide)](https://realpython.com/python-data-classes/)
- [Dan Bader: Writing Clean Python with Namedtuples](https://dbader.org/blog/writing-clean-python-with-namedtuples)
- [Trey Hunner: Easier Classes - Python Classes without All the Cruft](https://treyhunner.com/easier-classes/nbpy2018.html)
- [SO:  What are data classes and how are they different?](https://stackoverflow.com/questions/47955263/what-are-data-classes-and-how-are-they-different-from-common-classes)
- [DataClass vs NamedTuple vs Object: A Battle of Performance in Python](https://medium.com/@jacktator/dataclass-vs-namedtuple-vs-object-for-performance-optimization-in-python-691e234253b9) -- _warning, this is a subscription service, so not great for `links.json`_
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)

  

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  - You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: There is currently no issue or files for the concept.  **They are TBD.**

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
-`.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.
- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by fish-food89** [comment]
These are things I use daily at work :rofl: I might be able to contribute one way or another.

First idea.  
I would teach the named tuple concept using [`typing.NamedTuple`](https://docs.python.org/3/library/typing.html#typing.NamedTuple) instead of `collections.namedtuple()`. They are essentially the same things, but the typing out of a `NamedTuple` is much closer to that of a `@dataclass` decorated class. It is also much easier to read than objects created using `collections.namedtuple()`.

Sure you can do some nice tricks with `collections.namedtuple()` for creating some fun and wild dynamic classes defined in runtime using some configuration file, but that is much less needed in the professional world compared to clearly typed out classes in source files. I love dynamic things, but creating readable and maintainable software usually trumps the fun of creating "magical" dynamic classes.

So for the sake of teaching the concept and preparing people for a professional environment I would teach the subject using `typing.NamedTuple`. Just my two cents to the jar.

**Comment by BethanyG** [comment]
Hi @fish-food89 👋🏽 

> I would teach the named tuple concept using typing.NamedTuple instead of collections.namedtuple(). They are essentially the same things, but the typing out of a NamedTuple is much closer to that of a @dataclass decorated class. It is also much easier to read than objects created using collections.namedtuple().

Because type hinting and type aliasing are both "newer" and **_optional_** in Python (_the cpython reference implementation of the run time ignores type annotations_), I think that those things need to be in their own "branch" of the concept tree, and far enough down that a student understands that these are _**optional tools**_. 

They have their uses and abuses -- and it is not a requirement that they be used, and they should not be used everywhere.  Contrary to popular belief, they are in no way a requirement of "professional" python programming.  They **are** a very useful tool in some contexts, and terrible in others.  

So if anything, I think I'd **add** `typing.NamedTuple`, and go over the differences or rationales behind why all three types appear in Python.  Conversely, I might chuck mention of `namedtuples` altogether, and focus in on why `dataclasses` evolved in the first place.  Analogous to [PEP-557s](https://www.python.org/dev/peps/pep-0557/) original rationale (very small snips below).  

![image](https://user-images.githubusercontent.com/5923094/148701340-325681d4-75b2-4961-99d3-997006c0c39b.png)

`...`

![image](https://user-images.githubusercontent.com/5923094/148701833-e2095c38-9347-4c1b-b689-ba197d48c095.png)


But the idea of the exercise here (for me) was to build on what I'd consider the "basic" and "intermediate" data structures -- including classes -- and demonstrate that Python gives you multiple _flavors_ of container -- each with their own rationale or properties.  Not discuss mutability, immutability, nor any dynamic programming.

**Comment by fish-food89** [comment]
I see where you are coming from. And I agree in the benefit of introducing both versions of a "named tuple" to a student. But yeah, not necessarily all in the same concept exercise where Data Classes are being introduced. I think the amount of information needed to ingest with Data Classes alone is quite sufficient in itself. So if all three were to be covered in the concept tree I would give them all their own concept exercises. Start off with one of them, say Data Classes, and unlock the next exercises after completing that and have them build upon the things learnt from the Data Classes exercise. Or start off with one of the named tuples and build from there, doesn't really matter that much in what order you start IMO.

**Comment by BethanyG** [comment]
Closing in favor of the split up #3109  and #3110.
